cctp_primitives/src/commitment_tree/hashers.rs

```diff
@@ -1,152 +1,233 @@
+    withdrawal_epoch_length: u32,
+    cert_proving_system: u8,
+    csw_proving_system: u8,
+    mc_btr_request_data_length: u8,
+    custom_field_elements_configs: &[u8],
+    custom_bitvector_elements_configs: &[(u32, u32)],
+    btr_fee: u64,
+    ft_min_fee: u64,
+    // TODO: verify if it's enough to add to the comm_tree just the Poseidonhash of the custom_creation_data (Oleksandr)
+    custom_creation_data_hash: &[u8; FIELD_SIZE],
+    constant: Option<&[u8; FIELD_SIZE]>,
+    cert_verification_key_hash: &[u8; FIELD_SIZE],
+    csw_verification_key_hash: Option<&[u8; FIELD_SIZE]>
+) -> Result<FieldElement, Error>
+{
+    //TODO: Is there a reason why we deserialize tx_data_fes, sc_base_conf_fes and fees_fes separately ?
+    //      Can't we concatenate them in a single byte array and then deserialize it ?
+
+    // Init hash input
+    let mut fes = Vec::new();
+
+    // Convert tx data to field elements
+    let mut tx_data_fes = {
+        let mut buffer = Vec::new();
+        amount.write(&mut buffer)?;
+        pub_key.write(&mut buffer)?;
+        tx_hash.write(&mut buffer)?;
+        out_idx.write(&mut buffer)?;
+        bytes_to_field_elements(buffer)
+    }?;
+    fes.append(&mut tx_data_fes);
+
+    // Convert sc base configuration data into field elements
+    let mut sc_base_conf_fes = {
+        let mut buffer = Vec::new();
+        withdrawal_epoch_length.write(&mut buffer)?;
+        cert_proving_system.write(&mut buffer)?;
+        csw_proving_system.write(&mut buffer)?;
+        mc_btr_request_data_length.write(&mut buffer)?;
+        bytes_to_field_elements(buffer)
+    }?;
+    fes.append(&mut sc_base_conf_fes);
+
+
+    // Convert custom configuration data into field elements
+    let mut custom_conf_data_fes = {
+        let mut buffer = Vec::new();
+        custom_field_elements_configs.write(&mut buffer)?;
+        custom_bitvector_elements_configs.write(&mut buffer)?;
+        bytes_to_field_elements(buffer)
+    }?;
+    fes.append(&mut custom_conf_data_fes);
+
+    // Pack btr_fee and ft_min_fee into a single field element
+    fes.append(&mut bytes_to_field_elements(vec!([btr_fee, ft_min_fee]))?);
+
+    // Read the other data as field elements if present and push it to fes
+    fes.push(FieldElement::read(&custom_creation_data_hash[..])?);
+
+    if constant.is_some() { fes.push(FieldElement::read(&constant.unwrap()[..])?); }
+
+    fes.push(FieldElement::read(&cert_verification_key_hash[..])?);
+
+    if csw_verification_key_hash.is_some() { fes.push(FieldElement::read(&csw_verification_key_hash.unwrap()[..])?); }
+
+    // Compute final hash
+    Ok(hash_vec(fes))
 }
 
-// Converts byte-array into a sequence of FieldElements
-fn bytes_to_field_elements(bytes: &[u8]) -> Result<Vec<FieldElement>, Error> {
-    bytes_to_bits(bytes).to_field_elements()
+// Computes FieldElement-based hash on the given Ceased Sidechain Withdrawal data
+//TODO verify with Oleksandr if we need to add the redeemScript (signature?)
```

Comment:
I think - no. We don't use it on SC side and moreover SC trusts MC checks and can't check the signature by itself.